cmd/release-controller/sync_verify.go

```diff
@@ -56,6 +55,10 @@ func (c *Controller) ensureVerificationJobs(release *Release, releaseTag *imagev
 							continue
 						}
 					}
+					// Write previous run status into history before creating a new job
+					if len(status.History) == 0 || status.History[len(status.History)-1].URL != status.URL {
```

Comment:
Unfortunately this may not be safe because it will dramatically grow the size of our annotations and thus the size of the image stream.

For context, because each tag is annotations on an image stream we have to be conscious of the size of our image streams.

```
$ oc get is -n ocp release -o json | jq -c | wc -c
 1523015
$ oc get is -n ocp release -o json | jq -c '.spec' | wc -c
 1269693
$ oc get is -n ocp release -o json | jq -c '.status' | wc -c
  251877
```

We are dangerously close to the limit of object size allowed by etcd (~2mb limit but there are lots of factors at play).  So we have to be careful about the size we add.  If we only stored this on failing jobs, that would potentially keep the limit down, but we'd need to back of the envelope how much it would take.

Can you look at the ocp release image and estimate how much this would grow the size of that object, taking into account the average number of retries, etc?

Reply:
If we can figure out how to dramatically cut annotation size this becomes more reasonable.  If you can do the back of the envelope we can talk next week about how to improve it.

Reply:
Looks like transition time bloats the annotation by quite a bit.

Our current annotation size looks like:
```
$ oc get is -n ocp release -o json | jq -c '.spec.tags[].annotations' | wc -c
677761
```

Adding in the history, we get
```bash
$ oc get is -n ocp release -o json | jq -c '.spec.tags[].annotations' | \
	sed 's/\(\\"url\\":\\"\([^"]*\)\\",\\"retries\\":1\)/\1,\\"history\\":[{\\"url\\":\\"\2\\",\\"state\\":\\"Failed\\"}]/g' | \
	sed 's/\(\\"url\\":\\"\([^"]*\)\\",\\"retries\\":2\)/\1,\\"history\\":[{\\"url\\":\\"\2\\",\\"state\\":\\"Failed\\"},{\\"url\\":\\"\2\\",\\"state\\":\\"Failed\\"}]/g' | \
	sed 's/\(\\"url\\":\\"\([^"]*\)\\",\\"retries\\":3\)/\1,\\"history\\":[{\\"url\\":\\"\2\\",\\"state\\":\\"Failed\\"},{\\"url\\":\\"\2\\",\\"state\\":\\"Failed\\"},{\\"url\\":\\"\2\\",\\"state\\":\\"Failed\\"}]/g' | \
	sed 's/,\\"transitionTime\\":\\"[0-9]\{4\}-[0-9]\{2\}-[0-9]\{2\}T[0-9]\{2\}:[0-9]\{2\}:[0-9]\{2\}Z//g' | \
	wc -c
697002
```
This was for adding in 118 retries we have currently

If we can count on the base url for jobs to stay the same, we could get away with just storing the job numbers instead.
```bash
$ oc get is -n ocp release -o json | jq -c '.spec.tags[].annotations' | \
	sed 's/\(\\"url\\":\\"[^"]*\/\([^"\/]*\)\\",\\"retries\\":1\)/\1,\\"history\\":[{\\"url\\":\\"\2\\",\\"state\\":\\"Failed\\"}]/g' | \
	sed 's/\(\\"url\\":\\"[^"]*\/\([^"\/]*\)\\",\\"retries\\":2\)/\1,\\"history\\":[{\\"url\\":\\"\2\\",\\"state\\":\\"Failed\\"},{\\"url\\":\\"\2\\",\\"state\\":\\"Failed\\"}]/g' | \
	sed 's/\(\\"url\\":\\"[^"]*\/\([^"\/]*\)\\",\\"retries\\":3\)/\1,\\"history\\":[{\\"url\\":\\"\2\\",\\"state\\":\\"Failed\\"},{\\"url\\":\\"\2\\",\\"state\\":\\"Failed\\"},{\\"url\\":\\"\2\\",\\"state\\":\\"Failed\\"}]/g' | \
	wc -c
684351
```

We can remove transitionTime on jobs that don't have any more retries since that's just used to schedule the next retry
```bash
$ oc get is -n ocp release -o json | jq -c '.spec.tags[].annotations' | \
	sed 's/\(\\"url\\":\\"\([^"]*\)\\",\\"retries\\":1\)/\1,\\"history\\":[{\\"url\\":\\"\2\\",\\"state\\":\\"Failed\\"}]/g' | \
	sed 's/\(\\"url\\":\\"\([^"]*\)\\",\\"retries\\":2\)/\1,\\"history\\":[{\\"url\\":\\"\2\\",\\"state\\":\\"Failed\\"},{\\"url\\":\\"\2\\",\\"state\\":\\"Failed\\"}]/g' | \
	sed 's/\(\\"url\\":\\"\([^"]*\)\\",\\"retries\\":3\)/\1,\\"history\\":[{\\"url\\":\\"\2\\",\\"state\\":\\"Failed\\"},{\\"url\\":\\"\2\\",\\"state\\":\\"Failed\\"},{\\"url\\":\\"\2\\",\\"state\\":\\"Failed\\"}]/g' | \
	sed 's/,\\"transitionTime\\":\\"[0-9]\{4\}-[0-9]\{2\}-[0-9]\{2\}T[0-9]\{2\}:[0-9]\{2\}:[0-9]\{2\}Z//g' | \
	wc -c
614364
```

Storing past job numbers and removing transitionTime, we get some reduction in annotation size
```bash
$ oc get is -n ocp release -o json | jq -c '.spec.tags[].annotations' | \
	sed 's/\(\\"url\\":\\"[^"]*\/\([^"\/]*\)\\",\\"retries\\":1\)/\1,\\"history\\":[{\\"url\\":\\"\2\\",\\"state\\":\\"Failed\\"}]/g' | \
	sed 's/\(\\"url\\":\\"[^"]*\/\([^"\/]*\)\\",\\"retries\\":2\)/\1,\\"history\\":[{\\"url\\":\\"\2\\",\\"state\\":\\"Failed\\"},{\\"url\\":\\"\2\\",\\"state\\":\\"Failed\\"}]/g' | \
	sed 's/\(\\"url\\":\\"[^"]*\/\([^"\/]*\)\\",\\"retries\\":3\)/\1,\\"history\\":[{\\"url\\":\\"\2\\",\\"state\\":\\"Failed\\"},{\\"url\\":\\"\2\\",\\"state\\":\\"Failed\\"},{\\"url\\":\\"\2\\",\\"state\\":\\"Failed\\"}]/g' | \
	sed 's/,\\"transitionTime\\":\\"[0-9]\{4\}-[0-9]\{2\}-[0-9]\{2\}T[0-9]\{2\}:[0-9]\{2\}:[0-9]\{2\}Z//g' | \
	wc -c
602194
```

Reply:
Did you make the change for transition time in this PR or no (for non terminal)?

Reply:
Yes, it's removing transitionTime for terminal releases now. I could also add it in the sync ready phase so that transitionTime gets removed before a release is accepted/rejected